File: javascript/MyArrayList.js
class MyArrayList {
  constructor() {
    this.data = [];
    this.size = 0;
  }
  add(element) {
    this.data[this.size] = element;
    this.size++;
  }

  get(index) {
    if(index < 0 || index >= this.size) {
      throw new Error("index out of bounds");
    }
    return this.data[index];
  }

  remove(index) {
    if(index < 0 || index >= this.size) {
      throw new Error("index out of bounds");
    }
    for(let i = index; i < this.size - 1; i++) {
      this.data[i] = this.data[i + 1];
    }
    this.size--;
  }

  size() {
    return this.size;
  }

  isEmpty() {
    return this.size === 0;
  }
}

const myList = new MyArrayList();
myList.add("apple");
myList.add("banana");
myList.add("orange");

console.log(myList.get(1));
myList.remove(1);
console.log(myList.size());

File: java/MyArrayList.java
public class MyArrayList<E> {
  private static final int DEFAULT_CAPACITY = 10;

  private Object[] data;
  private int size;

  public MyArrayList() {
    data = new Object[DEFAULT_CAPACITY];
    size = 0;
  }

  public void add(E element) {
    if(size == data.length) {
      // resize array
      Object[] newData = new Object[data.length * 2];
      System.arraycopy(data, 0, newData, 0, data.length);
      data = newData;
    }
    data[size++] = element;
  }

  public E get(int index) {
    if(index < 0 || index >= size) {
      throw new IndexOutOfBoundsException();
    }
    return (E) data[index];
  }

  public void remove(int index) {
    if(index < 0 || index >= size) {
      throw new IndexOutOfBoundsException();
    }
    // shift elements to left
    System.arraycopy(data, index + 1, data, index, size - index - 1);
    size--;
  }

  public int size() {
    return size;
  }
  public boolean isEmpty() {
    return size == 0;
  }
}
